x/ccv/consumer/ibc_module_test.go

```diff
@@ -0,0 +1,374 @@
+					mocks.MockScopedKeeper.EXPECT().ClaimCapability(
+						ctx, params.chanCap, host.ChannelCapabilityPath(
+							params.portID, params.channelID)).Return(nil).Times(1),
+					mocks.MockConnectionKeeper.EXPECT().GetConnection(
+						ctx, "connectionIDToProvider").Return(
+						conntypes.ConnectionEnd{ClientId: "unexpectedClientID"}, true).Times(1), // unexpected clientID
+				)
+			}, false,
+		},
+	}
+
+	for _, tc := range testCases {
+
+		// Common setup
+		consumerKeeper, ctx, ctrl, mocks := testkeeper.GetConsumerKeeperAndCtx(
+			t, testkeeper.NewInMemKeeperParams(t))
+		consumerModule := consumer.NewAppModule(consumerKeeper)
+
+		consumerKeeper.SetPort(ctx, ccv.ConsumerPortID)
+		consumerKeeper.SetProviderClientID(ctx, "clientIDToProvider")
+
+		// Instantiate valid params as default. Individual test cases mutate these as needed.
+		params := params{
+			ctx:            ctx,
+			order:          channeltypes.ORDERED,
+			connectionHops: []string{"connectionIDToProvider"},
+			portID:         ccv.ConsumerPortID,
+			channelID:      "consumerChannelID",
+			chanCap:        &capabilitytypes.Capability{},
+			counterparty:   channeltypes.NewCounterparty(ccv.ProviderPortID, "providerChannelID"),
+			version:        ccv.Version,
+		}
+
+		tc.setup(ctx, &consumerKeeper, &params, mocks)
+
+		err := consumerModule.OnChanOpenInit(
+			params.ctx,
+			params.order,
+			params.connectionHops,
+			params.portID,
+			params.channelID,
+			params.chanCap,
+			params.counterparty,
+			params.version,
+		)
+
+		if tc.expPass {
+			require.NoError(t, err)
+		} else {
+			require.Error(t, err)
+		}
+		// Confirm there are no unexpected external keeper calls
+		ctrl.Finish()
+	}
+}
+
+// TestOnChanOpenTry validates the consumer's OnChanOpenTry implementation against the spec.
+//
+// See: https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/methods.md#ccv-ccf-cotry1
+// Spec tag: [CCV-CCF-COTRY.1]
+func TestOnChanOpenTry(t *testing.T) {
+
+	consumerKeeper, ctx, ctrl, _ := testkeeper.GetConsumerKeeperAndCtx(t, testkeeper.NewInMemKeeperParams(t))
+	// No external keeper methods should be called
+	defer ctrl.Finish()
+	consumerModule := consumer.NewAppModule(consumerKeeper)
+
+	// OnOpenTry must error even with correct arguments
+	_, err := consumerModule.OnChanOpenTry(
+		ctx,
+		channeltypes.ORDERED,
+		[]string{"connection-1"},
+		ccv.ConsumerPortID,
+		"channel-1",
+		nil,
+		channeltypes.NewCounterparty(ccv.ProviderPortID, "channel-1"),
+		ccv.Version,
+	)
+	require.Error(t, err, "OnChanOpenTry callback must error on consumer chain")
+}
+
+// TestOnChanOpenAck validates the consumer's OnChanOpenAck implementation against the spec.
+//
+// See: https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/methods.md#ccv-ccf-coack1
+// Spec tag: [CCV-CCF-COACK.1]
+func TestOnChanOpenAck(t *testing.T) {
+
+	// Params for the OnChanOpenAck method
+	type params struct {
+		ctx                   sdk.Context
```

Comment:
Why do you store a `ctx` copy here? Just asking bc both `ctx` and `params` are passed to the setup function.

Reply:
Nice catch and fixed, a duplicate `ctx` parameter is not needed for the setup function. `Params` is intended to match the parameters needed to call `OnChanOpenInit`

Reply:
@sainoe whoops looks like I only addressed your comments for one of the tests in this file, now both are fixed as of d569e93439de2cd42ef113cf8b028bf5d8b37c6b